### PR TITLE
[Live] Robust handling of invalid user data

### LINKED ISSUE #528: [RFC] Handling Data with Bad Types (i.e. that fails hydration)
Hi!

Consider this component:

```php
#[AsLiveComponent('counter')]
final class CounterComponent
{
    #[LiveProp(writable: true)]
    public int $count = 0;
}
```

In the template, it's bound to an input:

```
<input data-model="count">
```

What should happen if I enter `apple` into that box? (Note, yes we could use `<input type="number">`, but there are other examples , like a `DateTime` property where an invalid date time string is passed).

I can only think of two options:

A) We advise to NOT use property-types for writeable properties. The idea is that the component represents the "client-side data". And so if you can type `apple` into a field, then we should allow for `apple` to be entered on the property. Then, you use `#[Assert\....]` validation to validate things.

B) If there is a type mismatch, we reject the new value then, somewhere, expose an "error" that the user can render (e.g. `errors.get('count')` would have a message). 

(B) is attractive, but tricky. It would require:

1) On the Ajax call, in addition to the "latest data" (e.g. the latest `count` value), we would also need to send all of the "original data". This is needed because, if count was previously `10` and the user types in `apple`, we need to know the original value (`10`) so we can set it onto the property (since we will not be setting the string `apple`). If we didn't do this, the invalid property would go back to its original default value (if it has one, e.g. 0, which would look weird) or, worse, it would remain "uninitialized" and result in an error.

2) On the frontend, we would need to be smart enough to realize that if we send a value of `apple` for `count`, when that Ajax call finishes, because `count` was invalid, that data is still in an "unsynchronized" state. This means that the the `<input` field should *not* be updated to hold the latest value from the server (which would be the original `10` at this point).

So, solution (B) is doable. But we should be sure that it's not overkill - i.e. that solution (A) doesn't work well enough or something else.

### PR DESCRIPTION
**Pull request by weaverryan**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| Tickets       | Fix #528
| License       | MIT

Hi!

Another boring PR from me 😄.

t;dr When making an Ajax request, we now send the "original data" + "updated date". If the user sends something wacko (e.g. the word `apple` to a property with an `int` type), we can reject it and use the original data.

Cheers!

### CODE REVIEW

File: src/LiveComponent/src/Util/DehydratedProps.php [hunk 1]
```diff
@@ -0,0 +1,236 @@
+<?php
+
+namespace Symfony\UX\LiveComponent\Util;
+
+class DehydratedProps
```
Comment (kbond) [inline src/LiveComponent/src/Util/DehydratedProps.php:22]:
```suggestion
/**
 * Helps organize the dehydrated props and data.
 *
 * @author Ryan Weaver <ryan@symfonycasts.com>
 *
 * @experimental
 *
 * @internal
 */
class DehydratedProps
```

File: src/LiveComponent/src/Util/DehydratedProps.php [hunk 2]
```diff
@@ -0,0 +1,236 @@
+<?php
```
Comment (kbond) [inline src/LiveComponent/src/Util/DehydratedProps.php:1]:
```suggestion
<?php

/*
 * This file is part of the Symfony package.
 *
 * (c) Fabien Potencier <fabien@symfony.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
```

**Comment by norkunas** [comment]
As this affects hydration, I have a small proposition: add `dehydrationContext` to `AsLiveComponent` attribute which would be passed down to `serializer->denormalize` context.

Use case:
I have a root DTO for a form type and I declare on the DTO properties with docblock type hints and validation constraints, but without strict typing to avoid errors.

So now the result is that my initial DTO does not hold any data because it must be filled, and if I change a single field then the dehydration fails because the initial values are `null` and serializer gets from the property info that the value must be a string for example. (I know that I could initialize them as empty strings, so the type would match, but I can't initialize the prop if it expects an entity from an EntityType because it's not loaded at that time)

So my expectations that the form validation would handle that the value is not a string or an empty string, etc. and at the end I'd have a valid DTO which I could use.

Currently my workaround is to create a custom denormalizer and pass down `AbstractObjectNormalizer::DISABLE_TYPE_ENFORCEMENT => true` but that's not a friendly solution :)

**Comment by weaverryan** [comment]
I'm not sure specifically about your use-case... Regardless, you're 100% correct about needing a way to pass in the context for the serializer.

**Comment by weaverryan** [comment]
(de)normalization context added - usage for your situation

```php
#[LiveProp(
    denormalizationContext: [ObjectNormalizer::DISABLE_TYPE_ENFORCEMENT => true]
)]
public YourDto $yourDto;
```

**Comment by kbond** [comment]
What do you think about reusing the context attribute from serializer?

```php
#[LiveProp]
#[Context([ObjectNormalizer::DISABLE_TYPE_ENFORCEMENT => true])]
public YourDto $yourDto;
```

**Comment by norkunas** [comment]
> What do you think about reusing the context attribute from serializer?
> 
> ```php
> #[LiveProp]
> #[Context([ObjectNormalizer::DISABLE_TYPE_ENFORCEMENT => true])]
> public YourDto $yourDto;
> ```

Same thing to complete same task, I'd agree nevertheless ☺️

**Comment by kbond** [comment]
After I wrote that I was thinking that a property on LiveProp would be more "discoverable"

**Comment by weaverryan** [comment]
I don't think discoverability is very important as this is more of an edge-case feature. So I agree with using `#[Context]`. Btw, that feature (other than the "New in Symfony 5.3") looks undocumented - someone should fix that ;)

**Comment by weaverryan** [comment]
Updated to use `#[Context]`.

Note: while we use the serialization system, we only read the "context": we're not reading and using other options like "maxDepth", etc. Also, if you have `#[Context]` on a live prop, then that context is used to (de)normalize any writable paths. For example, if the `LiveProp` is a `Post` object and it has a writable path that is a `Category` object, we don't also look on *that* property to see if IT also has a `Context`. I want to avoid getting in the weeds with complex serialization edge-cases. And, so far, other thank @norkunas' use-case, nobody has even asked for control over the serialization context yet. So giving some basic control, I think, is enough.

**Comment by weaverryan** [comment]
Merging this - it's big and important - if there are any kinks, now we can work them out.

**Pull request merged**